### PR TITLE
[Infra] Add helper method to check if UI thread is required

### LINKED ISSUE #4854: [Infra] Add helper class to check whether the UI thread is required
### Description

The VS threading cookbook describes scenarios that should be  free-threaded and includes some sample code to check a particular set of operations is free-threaded or not.

TODO - add a helper class with documentation to make it easier to test whether code is free-threaded during development.

### PR DESCRIPTION
**Pull request by duncanp-sonar**
Fixes #4854

### CODE REVIEW

File: src/Infrastructure.VS/DEBUG_ONLY_VsCookbookThreadingChecker.cs [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+    // **
+    // **  Not be to used in shipping production code !
+    // **
+    // **  Only use for testing whether code is free-threaded
+    // **  in local dev environments.
+    // **
+    // ******************************************************
+
+
+    /* From the VS threading cookbook: https://github.com/microsoft/vs-threading/blob/main/doc/cookbook_vs.md
```
Comment (duncanp-sonar) [inline src/Infrastructure.VS/DEBUG_ONLY_VsCookbookThreadingChecker.cs:38]: Added helper methods based on the sample code in the VS thread cookbook.

File: src/Infrastructure.VS/DEBUG_ONLY_VsCookbookThreadingChecker.cs [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+        /// <summary>
+        /// If this method is called on the UI thread it will check
+        /// whether the operation _requires_ the UI thread.
+        /// If the method is called from a background thread it will just execute the operation.
+        /// </summary>
+        public static void IfOnUIThreadCheckIfRequiresMainThread(Action operationToTest)
+        {
+            if (ThreadHelper.CheckAccess())
+            {
+                Debug.WriteLine("On UI thread: testing operation...");
```
Comment (duncanp-sonar) [inline src/Infrastructure.VS/DEBUG_ONLY_VsCookbookThreadingChecker.cs:73]: I'm deliberately not using any of our other classes e.g. `ILogger` as they might skew the results.

File: src/Infrastructure.VS/DEBUG_ONLY_VsCookbookThreadingChecker.cs [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+
+    */
+    public static class DEBUG_ONLY_VsCookbookThreadingChecker
+    {
+        /// <summary>
+        /// If this method is called on the UI thread it will check
+        /// whether the operation _requires_ the UI thread.
+        /// If the method is called from a background thread it will just execute the operation.
+        /// </summary>
+        public static void IfOnUIThreadCheckIfRequiresMainThread(Action operationToTest)
```
Comment (duncanp-sonar) [inline src/Infrastructure.VS/DEBUG_ONLY_VsCookbookThreadingChecker.cs:69]: There are four methods:

* _IfOnUIThreadCheckIfRequiresMainThread(Action)_  : only does the check if the caller is on the UI thread.
* _CheckIfRequiresMainThread(Action)_  : does the check

... and equivalent methods for `Func<T>`.

**Review by bigfluffycookie — APPROVED** [review]
Looks good! we need to remember to fill in the others about this class so they can use it in the future! :)

**Pull request merged**